SpatialGDK/Source/SpatialGDK/Public/EngineClasses/Components/SpatialPingComponent.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "Components/ActorComponent.h"
+
+#include "SpatialPingComponent.generated.h"
+
+/*
+ Offers a configurable means of measuring round-trip latency in SpatialOS deployments.
+ This component should be attached to a player controller.
+ */
+UCLASS(ClassGroup = (SpatialGDK), Meta = (BlueprintSpawnableComponent))
+class SPATIALGDK_API USpatialPingComponent : public UActorComponent
+{
+	GENERATED_UCLASS_BODY()
+
+public:
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = SpatialPing)
+	bool bStartWithPingEnabled = true;
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = SpatialPing)
+	float PingFrequency = 0.5f;
+
+	virtual void BeginPlay() override;
+	virtual void EndPlay(const EEndPlayReason::Type EndPlayReason) override;
+
+	//Returns whether this component can ping
+	UFUNCTION(BlueprintCallable, Category = "SpatialGDK|Ping")
+	bool GetIsPingEnabled() const;
+
+	//Set whether this component can ping
+	UFUNCTION(BlueprintCallable, Category = "SpatialGDK|Ping")
+	void SetPingEnabled(bool bSetEnabled);
+
+	//Returns the latest raw round-trip ping value in seconds
+	UFUNCTION(BlueprintCallable, Category = "SpatialGDK|Ping")
+	float GetPing() const;
+
+private:
+	float RTPing;
+	float LastReceivedPingID;
+	UPROPERTY(ReplicatedUsing = OnRep_ReplicatedPingID)
+	float ReplicatedPingID;
+
+	bool bIsPingEnabled = false;
+
+	FTimerHandle PingTimerHandle;
+
+	APlayerController* OwningController;
+
+	void EnablePing();
+
+	void DisablePing();
+
+	UFUNCTION()
+	void TickPingComponent();
+
+	void SendNewPing();
+
+	UFUNCTION()
+	virtual void OnRep_ReplicatedPingID();
+
+	UFUNCTION(Server, Unreliable, WithValidation)
+	virtual void SendServerWorkerPingID(const float PingID);
```

Comment:
What round trip is this measuring exactly. There isn't a client RPC which means this isn't going over a symmetric channel.

Reply:
From my understanding there isn't currently an easy way to measure ping by normal means so this is measuring the round trip of a value sent via RPC and then replicated back to the sending client. This does mean that it's impacted by the replication tick rate so it's not perfect, but it provides a means to compare latency between native Unreal and Spatial networking.

Reply:
If what you want is actually a ping then the rpc and replication times might be very different. So you end up with a skewed result.

Reply:
Yeah, I think it might be unavoidable though from what I understand of the system at present. I'm looking at this more as a best estimate for the time being. I'd be happy to have a proper chat about it though if you have any suggestions. I grabbed this ticket partly as a learning opportunity so I expect there are better approaches that I'm missing.